common/namer/namer.go

```diff
@@ -1,57 +1,132 @@
-	return childNamer
 }
 
 func (n Namer) ResourceName(parts ...string) string {
 	if len(parts) == 0 {
 		panic("Resource name requires at least one part to generate name")
 	}
 
-	var resourceName string
-	if n.prefix != "" {
-		resourceName += n.prefix + nameSep
-	}
-
-	return resourceName + strings.Join(parts, nameSep)
+	return joinWithMaxLength(math.MaxInt, append(n.prefixes, parts...))
 }
 
-func (n Namer) DisplayName(parts ...pulumi.StringInput) pulumi.StringInput {
+func (n Namer) DisplayName(maxLen int, parts ...pulumi.StringInput) pulumi.StringInput {
 	var convertedParts []interface{}
 	for _, part := range parts {
 		convertedParts = append(convertedParts, part)
 	}
 	return pulumi.All(convertedParts...).ApplyT(func(args []interface{}) string {
-		strArgs := []string{n.ctx.Stack()}
+		strArgs := make([]string, 1, 1+len(n.prefixes)+len(args))
+		strArgs[0] = n.ctx.Stack()
+		strArgs = append(strArgs, n.prefixes...)
 		for _, arg := range args {
 			strArgs = append(strArgs, arg.(string))
 		}
-		return n.ResourceName(strArgs...)
+		return joinWithMaxLength(maxLen, strArgs)
 	}).(pulumi.StringOutput)
 }
+
+func joinWithMaxLength(maxLength int, tokens []string) string {
+	// Check if non-truncated concatenation fits inside maximum length
+	if lo.Sum(lo.Map(tokens, func(s string, _ int) int { return len(s) }))+(len(tokens)-1)*len(nameSep) <= maxLength {
+		return strings.Join(tokens, nameSep)
+	}
+
+	// If a truncation is needed, a hash will be needed
+	hasher := fnv.New64a()
+	for _, tok := range tokens {
+		_, _ = io.WriteString(hasher, tok)
+	}
+	fullhash := fmt.Sprintf("%016x", hasher.Sum64())
+
+	// Compute the size of the hash suffix that will be appended to the output
+	hash := fullhash
+	hashSize := maxLength * hashSizePercent / 100
+	if len(hash) > hashSize {
+		hash = hash[:hashSize]
+	} else {
+		hashSize = len(hash)
+	}
+
+	// For test purpose, we have an option to completely strip the output of the hash suffix
+	// At this point, `hashSize` is the size of the hash suffix without the dash separator
+	// -1 means that we want to also strip the dash
+	if noHash {
+		hashSize = -1
+	}
+
+	// If there’s so many tokens that truncating all of them to a single character string and keeping only the dash separators
+	// would exceed the maximum length, we cannot do anything better than returning only the hash.
+	if len(tokens)+(len(tokens))*len(nameSep)+hashSize > maxLength {
+		if len(fullhash) > maxLength {
+			return fullhash[:maxLength]
+		}
+		return fullhash
+	}
+
+	var sb strings.Builder
+
+	// Truncate all tokens in the same relative proportion
+	totalOutputSize := maxLength - len(tokens)*len(nameSep) - hashSize
+	totalInputSize := lo.Sum(lo.Map(tokens, func(s string, _ int) int { return len(s) }))
+	prevY := 0
+	X := 0
+	for _, tok := range tokens {
+		X += len(tok)
+		nextY := (X*totalOutputSize + totalInputSize/2) / totalInputSize
+		tokSize := nextY - prevY
+		prevY = nextY
+		sb.WriteString(tok[:tokSize])
```

Comment:
Can you explain how do you ensure that `tokSize` is less than `len(tok)`?

Reply:
As explained above, `tokSize` is a rounded value of `len(token) * totalOutputSize / totalInputSize` and, at this point of the function, we know that we have `totalOutputSize < totalInputSize`
The rest are details to ensure that integer rounding division error do not accumulate.

Reply:
Can you explain why you have `totalInputSize/2` in `nextY := (X*totalOutputSize + totalInputSize/2) / totalInputSize`?

Reply:
Looking at @ogaca-dd question, I can see as I also did not understand this formula. Could you split in "safeRoundedRatio" and then using `X*safeRoundedRatio`?

If I get it right, you are using `+totalInputSize/2` to make the ratio smaller, by increasing by 0.5 the computed `nextY`, as this equals
`round(X*totalOutputSize/totalInputSize + 0.5)` if we were using floats.

Reply:
Is there a reason to use `0.5` instead of another factor ?

Reply:
Thank you @ogaca-dd for asking the question !

Reply:
So…

Reply:
If we look at [the graph in the comment above](https://github.com/DataDog/test-infra-definitions/pull/240#discussion_r1261087617), we see that what we are trying to compute is the linear interpolation between points `(0, 0)` and `(totalInputSize, totalOutputSize)` at abscissa `X`.
It’s nothing more than a rule of 3.

So, what we a trying to compute is mathematically :
```
(math) Y = X × totalOutputSize / totalInputSize
```
Except that numbers of characters are integer, so, we have to find the closest integer to that value.
That is mathematically :
```
(math) Y = round(X × totalOutputSize / totalInputSize)
```

The problem is that the Go integer division doesn’t return the round value of the real division, but the quotient of the Euclidean division, which is the floor of the real division.
The output of the following Go expression:
```go
(GO) a / b
```
where `a` and `b` are `int` variables is:
```
(math) floor( a / b )
```

So, we are leveraging the following property:
```
(math) round( x ) = floor( x + ½ )
```

Except that, with integer computing, we cannot represent ½.
So, we do:
```
(math) Y
= round( X × totalOutputSize / totalInputSize )
= floor( X × totalOutputSize / totalInputSize + ½ )
= floor( (X × totalOutputSize + totalInputSize/2) / totalInputSize)
```
And this later expression can be computed in Go:
```go
Y := (X*totalOutputSize + totalInputSize/2) / totalInputSize
```

Reply:
Thank you for this clear explanation 💯 . Can you add a little comment in the code explaining that `totalInputSize/2` is used to have the floor value?

Reply:
As this naive algorithm wasn’t as trivial to understand as I initially expected, I think it’s better to move back to the optimal well-known solution: #252.
At least, explanations about Bresenham’s algorithm are easy to find.